common/net/bolt9.py

```diff
@@ -0,0 +1,185 @@
+from math import ceil
+
+# feature_name: odd_bit
+known_features = {
+    "option_data_loss_protect": 0,
+    "initial_routing_sync": 2,
+    "option_upfront_shutdown_script": 4,
+    "gossip_queries": 6,
+    "var_onion_optin": 8,
+    "gossip_queries_ex": 10,
+    "option_static_remotekey": 12,
+    "payment_secret": 14,
+    "basic_mpp": 16,
+    "option_support_large_channel": 18,
+    "option_anchor_outputs": 20,
+}
+
+# Reversed map -> odd_bit : feature_name
+known_odd_bits = {v: k for k, v in known_features.items()}
+
+
+def check_feature_name_bit_pair(name, bit):
+    """
+    Checks whether a given name and bit par match for known features.
+
+    Args:
+        name (:obj:`str`): the feature name.
+        bit (:obj:`int`): the bit position.
+
+    Returns:
+        :obj:`bool`: For known features, returns True if the pair match. For unknown features, returns True if the bit
+        is unknown.
+    """
+
+    if name in known_features:
+        # The pair match
+        return bit in [known_features[name], known_features[name] + 1]
+    else:
+        # The name and bit are unknown
+        return not (bit in known_features.values() or bit + 1 in known_features.values())
+
+
+class Feature:
+    """
+    Feature represents a feature bit.
+
+    Args:
+        bit (:obj:`int`): the bit this feature holds in the feature vector.
+        is_set (:obj:`bool`): whether the feature is set or not.
+
+    Attributes:
+        is_odd (:obj:`bool`): whether the bit is odd or even.
+    """
+
+    def __init__(self, bit, is_set):
+        if not isinstance(bit, int):
+            raise TypeError("bit must be int")
+        if not isinstance(is_set, bool):
+            raise TypeError("is_set must be bool")
+
+        self.bit = bit
+        self.is_set = is_set
+        self.is_odd = bool(self.bit % 2)
+
+
+class FeatureVector:
+    """The FeatureVector encapsulated all the features."""
+
+    def __init__(self, **kwargs):
+        self._features = {}
+        for key, value in kwargs.items():
+            if not isinstance(value, Feature):
+                raise TypeError(f"Features must be of type Feature, {type(value)} received")
+            elif key == "initial_routing_sync" and value.is_set and not value.is_odd:
+                raise ValueError("initial_routing_sync has no even bit")
+            elif not check_feature_name_bit_pair(key, value.bit):
+                raise ValueError("Feature name and bit do not match")
+
+            vars(self)[key] = value
```

Comment:
This is equivalent to `self.__dict__[key] = value` I guess? Just wondering why `vars()` is used here.

Reply:
An alternative would be to have a custom `__getattr__` that first looks up into `self._features` and then calls the superclass `__getattr__`. That way the `self._features` dict remains authoritative and there's no chance for the two to be out of sync.